lightningd/htlc_set.c

```diff
@@ -0,0 +1,193 @@
+#include <common/memleak.h>
+#include <common/timeout.h>
+#include <lightningd/htlc_end.h>
+#include <lightningd/htlc_set.h>
+#include <lightningd/lightningd.h>
+#include <lightningd/peer_htlcs.h>
+
+/* Pre-declaration as fail_htlc_set refers to it and vice versa */
+static void htlc_set_hin_destroyed(struct htlc_in *hin,
+				   struct htlc_set *set);
+
+void fail_htlc_set(struct htlc_set *set, enum onion_type failcode)
+{
+	for (size_t i = 0; i < tal_count(set->htlcs); i++) {
+		/* Don't remove from set */
+		tal_del_destructor2(set->htlcs[i], htlc_set_hin_destroyed, set);
+		fail_htlc(set->htlcs[i], failcode);
+	}
+	tal_free(set);
+}
+
+/* If an HTLC times out, we need to free entire set, since we could be processing
+ * it in invoice.c right now. */
+static void htlc_set_hin_destroyed(struct htlc_in *hin,
+				   struct htlc_set *set)
+{
+	/* For non-EXPERIMENTAL_FEATURES, this never actually called. */
+#if EXPERIMENTAL_FEATURES
+	for (size_t i = 0; i < tal_count(set->htlcs); i++) {
+		if (set->htlcs[i] == hin) {
+			/* Don't try to re-fail this HTLC! */
+			tal_arr_remove(&set->htlcs, i);
+			/* Kind of the correct failure code. */
+			fail_htlc_set(set, WIRE_MPP_TIMEOUT);
+			return;
+		}
+	}
+	abort();
+#endif
+}
+
+void fulfill_htlc_set(struct htlc_set *set, const struct preimage *preimage)
+{
+	for (size_t i = 0; i < tal_count(set->htlcs); i++) {
+		/* Don't remove from set */
+		tal_del_destructor2(set->htlcs[i], htlc_set_hin_destroyed, set);
+		fulfill_htlc(set->htlcs[i], preimage);
+	}
+	tal_free(set);
+}
+
+#if EXPERIMENTAL_FEATURES
+static void destroy_htlc_set(struct htlc_set *set,
+			     struct htlc_set_map *map)
+{
+	htlc_set_map_del(map, set);
+}
+
+/* BOLT-9441a66faad63edc8cd89860b22fbf24a86f0dcd #4:
+ * - MUST fail all HTLCs in the HTLC set after some reasonable
+ *   timeout.
+ *   - SHOULD use `mpp_timeout` for the failure message.
+ */
+static void timeout_htlc_set(struct htlc_set *set)
+{
+	fail_htlc_set(set, WIRE_MPP_TIMEOUT);
+}
+#endif /* EXPERIMENTAL_FEATURES */
+
+static struct htlc_set *new_htlc_set(struct lightningd *ld,
+				     struct htlc_in *hin,
+				     struct amount_msat total_msat)
+{
+	struct htlc_set *set;
+
+	set = tal(ld, struct htlc_set);
+	set->total_msat = set->so_far = total_msat;
+	set->payment_hash = hin->payment_hash;
+	set->so_far = AMOUNT_MSAT(0);
+	set->htlcs = tal_arr(set, struct htlc_in *, 1);
+	set->htlcs[0] = hin;
+
+#if EXPERIMENTAL_FEATURES
+	/* BOLT-9441a66faad63edc8cd89860b22fbf24a86f0dcd #4:
+	 * - MUST fail all HTLCs in the HTLC set after some reasonable
+	 *   timeout.
+	 *   - SHOULD wait for at least 60 seconds after the initial
+	 *     HTLC.
+	 */
+	notleak(new_reltimer(ld->timers, set, time_from_sec(70),
+			     timeout_htlc_set, set));
+	htlc_set_map_add(&ld->htlc_sets, set);
+	tal_add_destructor2(set, destroy_htlc_set, &ld->htlc_sets);
+#endif
+	return set;
+}
+
+void htlc_add_set(struct lightningd *ld,
+		  struct htlc_in *hin,
+		  struct amount_msat total_msat,
```

Comment:
Nit: `total_msat` -> `msatoshi_total` to match other occurrences.

Reply:
total_msat is from spec, rest should be changed instead?